### PR TITLE
rpcclient: safe read and write to batch

### PR DESCRIPTION
**Pull request by ynewmann**
This PR tries to make batchList concurrent-safe.

The `Send()` could panic at the line 1730 if c.removeRequest  returned `nil`. It could return `nil` if we try to call `Send()` concurrently.
https://github.com/btcsuite/btcd/blob/684d64ad74fed203fb846c032f2b55b3e3c36734/rpcclient/infrastructure.go#L1708-L1730

After the quick look I've noticed that there are places where the batchList is update unsafely. So, the simplest fix is just to add locks, but I'm afraid if it could break something else, because I'm not very familiar with the library.

### CODE REVIEW

**Comment by Roasbeef** [comment]
Approved CI.

File: rpcclient/infrastructure.go [hunk 1]
```diff
@@ -1656,7 +1665,15 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
 	return c.backendVersion, nil
 }
 
-func (c *Client) sendAsync() FutureGetBulkResult {
+func (c *Client) sendAsync() (FutureGetBulkResult, error) {
+	c.batchLock.Lock()
+	defer c.batchLock.Unlock()
+
+	// if batchlist is empty, there's nothing to send
```
Comment (guggero) [inline rpcclient/infrastructure.go:1672]: nit (pre-existing from copy/paste below): Please use full sentences for comments (with initial capitalization and punctuation). Also, it should either say "batch list" or "batchList".

File: rpcclient/infrastructure.go [hunk 2]
```diff
@@ -1678,25 +1695,28 @@ func (c *Client) sendAsync() FutureGetBulkResult {
...
-	batchResp, err := c.sendAsync().Receive()
+	batchResp, err := future.Receive()
 	if err != nil {
 		// Clear batchlist in case of an error.
 		//
 		// TODO(yy): need to double check to make sure there's no
 		// concurrent access to this batch list, otherwise we may miss
 		// some batched requests.
+
+		c.batchLock.Lock()
```
Comment (guggero) [inline rpcclient/infrastructure.go:1717]: nit: can now remove the TODO above, as that's what the PR does.

**Review by guggero — COMMENTED** [review]
Thanks for the fix! Just two minor nits.

**Comment by ynewmann** [comment]
> Thanks for the fix! Just two minor nits.

One more nit here [b3d9947](https://github.com/btcsuite/btcd/pull/2273/commits/b3d9947a3706c83372ca94007b69c13eec20439e)

**Review by guggero — APPROVED** [review]
Can you squash the commits into one please?
Then this is good to go.

**Review by guggero — APPROVED** [review]

**Comment by yyforyongyu** [comment]
Could you do a rebase on master? I'm trying to test it in `lnd` but the current branch doesn't include some recent fixes that were pushed to master.

**Comment by ynewmann** [comment]
> Could you do a rebase on master? I'm trying to test it in `lnd` but the current branch doesn't include some recent fixes that were pushed to master.

done

**Review by yyforyongyu — APPROVED** [review]
LGTM 🚢

**Pull request merged**